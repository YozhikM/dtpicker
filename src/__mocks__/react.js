const react = require('react');
// Resolution for requestAnimationFrame not supported in jest error :
// https://github.com/facebook/react/issues/9102#issuecomment-283873039
global.window = global;
window.addEventListener = () => {};
window.requestAnimationFrame = () => {
  throw new Error('requestAnimationFrame is not supported in Node');
};

module.exports = react;

global.spyMount = (component, ...methodNames) => {
  const mount = require('enzyme').mount;
  const wrapper = mount(component);
  const result = [wrapper];
  methodNames.forEach(methodName => {
    result.push(jest.spyOn(wrapper.instance(), methodName));
  });
  wrapper.instance().forceUpdate();
  return result;
};
