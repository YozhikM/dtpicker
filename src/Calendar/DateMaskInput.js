/* @flow */

import * as React from 'react';
import { format } from 'date-fns';
import Input from '../Input/Input';
import type { Highlight } from './helpers';

type Props = {|
  visibleDate: ?Highlight,
  minDate?: Date,
  maxDate?: Date,
  onChange?: Function,
  onFocus?: Function,
  onKeyPress?: Function,
  style?: Object
|};

type State = {
  inputValue: string
};

export default class DateMaskInput extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);

    this.state = {
      inputValue: this.formatPropsToState(props.visibleDate)
    };
  }

  componentWillReceiveProps(nextProps: Props) {
    if (nextProps.visibleDate) {
      this.setState({
        inputValue: this.formatPropsToState(nextProps.visibleDate)
      });
    }
  }

  formatPropsToState = (value: ?Highlight) => {
    if (!value) return '';

    const date = new Date(Date.UTC(value.year, value.month, value.day));
    return this.formatWithoutTime(date);
  };

  formatWithTime(date: ?Date): string {
    return format(date, 'DD/MM/YYYY HH:mm');
  }

  formatWithoutTime(date: ?Date): string {
    return format(date, 'DD/MM/YYYY');
  }

  replaceText(text: string, rule: RegExp): string {
    if (text.match(rule)) return text.replace(rule, '');
    return text;
  }

  replaceValue(value: string): string {
    let copyValue = value.slice();
    const regArr = [/\//g, /\./g, /-/g];
    for (let i = 0; i < regArr.length; i++) {
      copyValue = this.replaceText(copyValue, regArr[i]);
    }

    return copyValue;
  }

  onChangeInputValue = (e: SyntheticInputEvent<>, value: string) => {
    const { onChange, minDate, maxDate } = this.props;

    let userDate;
    if (value.length > 10) {
      const splicedValue = value.slice(0, 10);
      this.setState({ inputValue: splicedValue });
    }

    const replacedValue = this.replaceValue(value);

    if (value.length === 10) {
      const YY = Number(value.substr(6, 4));
      const MM = Number(value.substr(3, 2)) - 1;
      const DD = Number(value.substr(0, 2));

      userDate = new Date(Date.UTC(YY, MM, DD));

      if (maxDate && minDate) {
        if (userDate < minDate) {
          this.setState({ inputValue: this.formatWithoutTime(minDate) });
          return;
        } else if (userDate > maxDate) {
          this.setState({ inputValue: this.formatWithoutTime(maxDate) });
          return;
        }
      }

      if (!maxDate && !minDate) {
        this.setState({ inputValue: this.formatWithoutTime(userDate) });
      }
    } else if (replacedValue.length === 8) {
      const YY = Number(replacedValue.substr(4, 4));
      const MM = Number(replacedValue.substr(2, 2)) - 1;
      const DD = Number(replacedValue.substr(0, 2));

      userDate = new Date(Date.UTC(YY, MM, DD));

      if (maxDate && minDate) {
        if (userDate < minDate) {
          this.setState({ inputValue: this.formatWithoutTime(minDate) });
          return;
        } else if (userDate > maxDate) {
          this.setState({ inputValue: this.formatWithoutTime(maxDate) });
          return;
        }
      }

      if (!maxDate && !minDate) {
        this.setState({ inputValue: this.formatWithoutTime(userDate) });
      }
    }

    if (userDate && onChange) {
      if ((maxDate && userDate <= maxDate) || (minDate && userDate >= minDate)) {
        onChange(userDate);
      } else if (maxDate && userDate > maxDate) {
        onChange(maxDate);
      } else if (minDate && userDate < minDate) {
        onChange(minDate);
      } else if (!maxDate && !minDate) {
        onChange(userDate);
      }
    }
  };

  render() {
    const { onFocus, onKeyPress, style } = this.props;
    const { inputValue } = this.state;

    return (
      <Input
        value={inputValue}
        onChange={this.onChangeInputValue}
        onFocus={onFocus}
        onKeyPress={onKeyPress}
        style={style}
        centerAlignedText
        small
      />
    );
  }
}
