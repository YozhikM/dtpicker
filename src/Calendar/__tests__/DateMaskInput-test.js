/* @flow */

import * as React from 'react';
import DateMaskInput from '../DateMaskInput';

const visibleDate = new Date('2018, 2, 9');

describe('DateMaskInput', () => {
  it('render()', () => {
    const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('componentWillReceiveProps()', () => {
    const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} />);
    const nextProps = {
      visibleDate: new Date('2017, 2, 10'),
    };
    wrapper.instance().componentWillReceiveProps(nextProps);

    expect(wrapper.state('inputValue')).toBe('10/02/2017');
  });

  it('formatWithTime()', () => {
    const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} />);
    const formattedDate = wrapper.instance().formatWithTime(new Date('2018, 2, 6'));

    expect(formattedDate).toBe('06/02/2018 00:00');
  });

  it('formatWithoutTime()', () => {
    const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} />);
    const formattedDate = wrapper.instance().formatWithoutTime(new Date('2018, 2, 1'));

    expect(formattedDate).toBe('01/02/2018');
  });

  it('replaceText()', () => {
    const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} />);
    const replacedText = wrapper.instance().replaceText('Маша', /ма/gi);

    expect(replacedText).toBe('ша');
  });

  it('replaceValue()', () => {
    const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} />);
    const replacedValue = wrapper.instance().replaceValue('20.03-2018/');

    expect(replacedValue).toBe('20032018');
  });

  describe('onChangeInputValue', () => {
    const onChange = jest.fn();

    it('if value > 10', () => {
      const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} onChange={onChange} />);
      wrapper.instance().onChangeInputValue(({}: any), '01/07/19931');

      expect(wrapper.state('inputValue')).toBe('01/07/1993');
      expect(onChange).not.toBeCalled();
    });

    it('if value === 10', () => {
      const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} onChange={onChange} />);
      wrapper.instance().onChangeInputValue(({}: any), '01/07/1994');

      expect(wrapper.state('inputValue')).toBe('01/07/1994');
      expect(onChange).toBeCalledWith(new Date('1994-06-30T17:00:00.000Z'));
    });

    it('replacedValue.length === 8', () => {
      const wrapper = shallow(<DateMaskInput visibleDate={visibleDate} onChange={onChange} />);
      wrapper.instance().onChangeInputValue(({}: any), '01091994');

      expect(wrapper.state('inputValue')).toBe('01/09/1994');
      expect(onChange).toBeCalledWith(new Date('1994-08-31T17:00:00.000Z'));
    });
  });
});
