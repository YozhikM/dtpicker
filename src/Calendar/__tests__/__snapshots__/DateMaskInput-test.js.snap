// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateMaskInput render() 1`] = `
<Input
  centerAlignedText={true}
  onChange={[Function]}
  small={true}
  value="09/02/2018"
/>
`;
